[Feat] Introduce shareable diffusion workflows

sayakpaul opened the pull request:
Fixes https://github.com/huggingface/diffusers/issues/4734. 

## Use cases

Beautifully laid out by @patrickvonplaten in https://github.com/huggingface/diffusers/pull/4825#issuecomment-1764728963.

1.) I want to create a reproducible workflow:

```python
from diffusers import DiffusionPipeline
import torch

pipeline = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, safety_checker=None)
pipeline.to("cuda")

outputs = pipeline("A painting of a horse", num_inference_steps=15, return_workflow=True)

workflow = outputs.workflow
```

Printing out `workflow` gives:

```
{'prompt': 'A painting of a horse',
 'height': 512,
 'width': 512,
 'num_inference_steps': 15,
 'guidance_scale': 7.5,
 'negative_prompt': None,
 'num_images_per_prompt': 1,
 'eta': 0.0,
 'output_type': 'pil',
 'return_dict': True,
 'callback': None,
 'callback_steps': 1,
 'cross_attention_kwargs': None,
 'guidance_rescale': 0.0,
 'clip_skip': None,
 '_name_or_path': 'runwayml/stable-diffusion-v1-5'}
```

2. I want to then share the workflow with others:

```python
outputs.workflow.push_to_hub("workflow-test")
```

3.1. I want to load the workflow:

```python
from diffusers import DiffusionPipeline
# from functools import partial
import torch

pipeline = DiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16
).to("cuda")

pipeline.load_workflow("sayakpaul/workflow-test")
image = pipeline().images[0]
```

3.2. I want to load the workflow with a specific filename:

```python
from diffusers import DiffusionPipeline
import torch

pipeline = DiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16
).to("cuda")

pipeline.load_workflow("sayakpaul/workflow-test", filename="horse_workflow.json")
image = pipeline().images[0]
```

4. I want to load a workflow saved locally:

```python
from diffusers import DiffusionPipeline
from diffusers.workflow_utils import _NON_CALL_ARGUMENTS
import torch
import json

pipeline = DiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16
).to("cuda")

with open("workflow-test/horse_workflow.json") as f:
    workflow = json.load(f)

workflow = {k: v for k, v in workflow.items() if k not in _NON_CALL_ARGUMENTS}
images = pipeline(**workflow).images[0]
```

The following could potentially be wrapped in a utility method and exposed from the `workflow_utils` module:

```python
with open("workflow-test/horse_workflow.json") as f:
    workflow = json.load(f)

workflow = {k: v for k, v in workflow.items() if k not in _NON_CALL_ARGUMENTS}
```

## Limitations

Workflows are meant to be simple and easily shareable. This means we don't serialize anything of the type `torch.Tensor`, `np.ndarray`, `PIL.Image.Image`, and `Callable`. So, `prompt_embeds` won't work, for example.  

Then, we need to consider how we want to populate information about LoRAs. C

## TODOs

- [ ] Propagate the changes to the other pipelines 
- [x] Add docs
- [x] Add tests

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_4825). All of your documentation changes will be reflected on that endpoint.

sayakpaul commented:
Actually, thinking a bit more about the design. 

I think it's better if we don't allow the workflow loading API to modify the `__call__()` argument and instead make it return the call arguments, which the user can pass to `pipe`. Something like:

```python
call_args = pipeline.load_from_workflow("workflow-path")
images = pipeline(**call_args).images
```

This way, we don't have to modify the call arguments dynamically and store them internally for other complementary APIs like `unload_workflow()`. Reduces the maintenance burden and does the job while staying clean and consistent IMO.

patrickvonplaten commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -1746,3 +1748,25 @@ def set_attention_slice(self, slice_size: Optional[int]):
         for module in modules:
             module.set_attention_slice(slice_size)
+
+    def load_workflow(self, workflow_id, workflow_filename=None):
```
```suggestion
    def load_workflow(self, workflow_id_or_path, workflow_filename=None):
```

patrickvonplaten commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -1746,3 +1748,25 @@ def set_attention_slice(self, slice_size: Optional[int]):
+        workflow_filename = workflow_filename or WORKFLOW_NAME
+
+        if os.path.isdir(workflow_id):
+            workflow_filepath = os.path.join(workflow_id, workflow_filename)
```
```suggestion
            workflow_filepath = os.path.join(workflow_id, workflow_filename)
        elif os.path.isfile(workflow_id_or_path):
            workflow_file_path = workflow_id_or_path
```

patrickvonplaten replied:
just to also allow to directly load from a Path

patrickvonplaten commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -1746,3 +1748,25 @@ def set_attention_slice(self, slice_size: Optional[int]):
+        self._update_call(**pipeline_call_args)
+
+    def _update_call(self, **kwargs):
+        self.__call__ = partial(self.__call__, **kwargs)
```
Cool idea! I slightly prefer to instead overwrite our current default values though, so that a user could e.g. load a workflow, but still overwrite values from the workflow. 

Could you maybe see if logic similar to this example could work well here:

```py
import inspect

# Define the original function with default values
def my_function(a=1, b=2, c=3):
    return a + b + c

# New default values you want to set
new_defaults = {
    'a': 10,
    'b': 20,
    'c': 30
}

# Get the original function's signature
original_signature = inspect.signature(my_function)

# Update the default values in the signature
new_params = []
for param_name, param in original_signature.parameters.items():
    if param_name in new_defaults:
        new_params.append(param.replace(default=new_defaults[param_name]))
    else:
        new_params.append(param)

# Create a new signature with modified default values
new_signature = original_signature.replace(parameters=new_params)

# Create a new function with the modified signature
def new_function(*args, **kwargs):
    bound_args = new_signature.bind(*args, **kwargs)
    bound_args.apply_defaults()
    return my_function(*bound_args.args, **bound_args.kwargs)

# Test the new function
result = new_function()
print(result)  # Output: 60 (10 + 20 + 30)
```

?

sayakpaul replied:
This is what I would prefer: https://github.com/huggingface/diffusers/pull/4825#issuecomment-1697304207.

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -718,7 +722,14 @@ def __call__(
+
         if not return_dict:
-            return (image, has_nsfw_concept)
+            return (image, has_nsfw_concept, workflow)
```
```suggestion
            outputs = (image, has_nsfw_concept)
            
            if workflow is not None:
                outputs += (workflow,)
                
            return outputs
```

sayakpaul replied:
Better. Thanks!

patrickvonplaten commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -1746,3 +1748,25 @@ def set_attention_slice(self, slice_size: Optional[int]):
+            workflow_filepath = os.path.join(workflow_id, workflow_filename)
+        else:
+            workflow_filepath = hf_hub_download(repo_id=workflow_id, filename=workflow_filename)
+        workflow = self._dict_from_json_file(workflow_filepath)
```
```suggestion
        self.workflow = self._dict_from_json_file(workflow_filepath)
```
I'd just save the workflow as an attribute here no?

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -718,7 +722,14 @@ def __call__(
+        if return_workflow:
+            signature = inspect.signature(self.__call__)
+            argument_names = [param.name for param in signature.parameters.values()]
+            call_arg_values = inspect.getargvalues(inspect.currentframe()).locals
```
nice! does that also get the kwargs values?

sayakpaul replied:
If so, yes, it does.

patrickvonplaten reviewed (commented):
Cool first draft - will need to play around with it a bit more! But generally I do like the design :-)

patrickvonplaten commented on src/diffusers/workflow_utils.py:
```diff
@@ -0,0 +1,88 @@
+    return workflow
+
+
+class Workflow(ConfigMixin, PushToHubMixin):
```
Let's keep all workflows to just be json dicts, no need to introduce new concepts here that the user needs to understand

patrickvonplaten commented on src/diffusers/workflow_utils.py:
```diff
@@ -0,0 +1,88 @@
+        workflow["lora"].update(lora_info)
+
+    # Populate component configs.
+    workflow.update(
+        {
+            "components": {
+                component_name: component.config
+                if isinstance(component.config, FrozenDict)
+                else component.config.to_dict()
+                for component_name, component in pipeline_components.items()
+                if hasattr(component, "config")
+            }
+        }
+    )
```
```suggestion
```

This is not part of the workflow IMO. The workflow (by naming) is really only the parameters passed to the __call__ method by the user. The model's config is already nicely stored in a config file

sayakpaul replied:
Yeah I agree.

sayakpaul commented:
@patrickvonplaten thanks for the thoughtful review here. 

Almost everything has been addressed except for this one:

> and then the pipeline should be runnable without having to pass anything to the pipeline:

Let me elaborate on the problem with an example. Consider the following:

```python
class DummyClass:
    def load_values(self, prompt, num_inference_steps):
        self = partial(self, prompt=prompt, num_inference_steps=num_inference_steps)

class AnotherDummyClass(DummyClass):
    def __init__(self):
        self.some_val = 5

    def __call__(self, prompt=None, num_inference_steps=10):
        if prompt is None:
            raise ValueError("prompt cannot be None.")
        return f"Called with {prompt} and {num_inference_steps}"
```
Now, if you do:

```python
some_class = AnotherDummyClass()
some_class.load_values(prompt="Sayak", num_inference_steps=25)
some_class()
```

It leads to:

```
ValueError: prompt cannot be None.
```

_(patching `__cal__()` inside of `load_values()` also doesn't work)_

However, if you do:

```python
from functools import partial 

some_class = AnotherDummyClass()
some_class = partial(some_class, prompt="Sayak")
```

everything works as expected. 

This is exactly what we're facing at the moment. One way to remedy this would be to make `load_workflow()` return the call arguments and then use pipeline with `pipeline(**call_args_from_workflow)`. 

I also tried repurposing what you proposed in https://github.com/huggingface/diffusers/pull/4825#discussion_r1309811630 but didn't solve the problem. 

Thoughts?

sayakpaul commented:
@patrickvonplaten thanks for the thoughtful comment! 

I have addressed your concerns and also modified the PR comment to include the current use cases, limitations, and TODOs. I think the PR is ready for another review. Once it looks good, I can take care of the remaining TODOs.

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_output.py:
```diff
@@ -19,10 +19,13 @@ class StableDiffusionPipelineOutput(BaseOutput):
 
     images: Union[List[PIL.Image.Image], np.ndarray]
     nsfw_content_detected: Optional[List[bool]]
+    workflow: dict
```
this should be optional I think

patrickvonplaten commented on src/diffusers/configuration_utils.py:
```diff
@@ -157,7 +157,12 @@ def save_config(self, save_directory: Union[str, os.PathLike], push_to_hub: bool
 
         # If we save using the predefined names, we can load using `from_config`
-        output_config_file = os.path.join(save_directory, self.config_name)
+        filename = kwargs.pop("filename", None)
```
Uff not too happy about such fundamental changes. Why is this needed?

sayakpaul replied:
Lets us pass a filename to save a workflow file. Users can always manually rename files.

patrickvonplaten commented on src/diffusers/loaders.py:
```diff
@@ -1168,6 +1168,8 @@ class LoraLoaderMixin:
     """
     text_encoder_name = TEXT_ENCODER_NAME
     unet_name = UNET_NAME
+    loras_loaded = 0
```
would we do this maybe in a v2 - I'd prefer to keep the v1 version as simple as possible and store only __call__ arguments

sayakpaul replied:
Yeah agreed.

patrickvonplaten commented on src/diffusers/loaders.py:
```diff
@@ -1224,6 +1226,11 @@ def load_lora_weights(
             adapter_name=adapter_name,
             _pipeline=self,
         )
+        if not USE_PEFT_BACKEND:
```
would be great to do this maybe in a follow up PR

sayakpaul replied:
Yeah in v2, preferably.

patrickvonplaten commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -1977,3 +1981,52 @@ def set_attention_slice(self, slice_size: Optional[int]):
+            generator = torch.manual_seed(seed)
+        else:
+            generator = None
+        workflow.update({"generator": generator})
```
Can we serialize a `generator`?

sayakpaul replied:
We're not serializing a generator. See here:
https://github.com/huggingface/diffusers/blob/feat/workflows/src/diffusers/workflow_utils.py#L62C1-L67C44

DN6 replied:
Wondering if you'd also like to add info on the generator device.

sayakpaul replied:
Done.

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -779,7 +783,21 @@ def __call__(
 
+        workflow = None
+        if return_workflow:
+            signature = inspect.signature(self.__call__)
+            argument_names = [param.name for param in signature.parameters.values()]
+            call_arg_values = inspect.getargvalues(inspect.currentframe()).locals
+            workflow = populate_workflow_from_pipeline(
+                argument_names, call_arg_values, self.lora_info, self.config._name_or_path
+            )
```
```suggestion
            signature = inspect.signature(self.__call__)
            argument_names = [param.name for param in signature.parameters.values()]
            call_arg_values = inspect.getargvalues(inspect.currentframe()).locals
            workflow = populate_workflow_from_pipeline(
                argument_names, call_arg_values, self.lora_info, self.config._name_or_path
            )
```
should we move this to `pipeline_utils`?

sayakpaul replied:
We can but the problem is the scope. If you delegate this to another module, the scope gets disturbed and we no longer get access to the right values.

patrickvonplaten commented on src/diffusers/workflow_utils.py:
```diff
@@ -0,0 +1,104 @@
+    return workflow
+
+
+class Workflow(dict, ConfigMixin, PushToHubMixin):
```
Nice!

patrickvonplaten commented on src/diffusers/workflow_utils.py:
```diff
@@ -0,0 +1,97 @@
+
+
+class Workflow(dict, ConfigMixin, PushToHubMixin):
+    """Class sub-classing from native Python dict to have support for interacting with the Hub."""
```
```suggestion
    """Class sub-classing from native Python dict to have support for interacting with the Hub."""
    
    config_name = None
```

patrickvonplaten commented on src/diffusers/workflow_utils.py:
```diff
@@ -0,0 +1,97 @@
+        self._internal_dict.pop(__key)
+        super().pop(__key)
+
+    def save_workflow(self, **kwargs):
```
```suggestion
    def save_workflow(self, filename: str):
        self.config_name = filename
```

patrickvonplaten replied:
Can we try to avoid using `kwargs` here and try to write out all the function signature arguments?

patrickvonplaten commented on src/diffusers/workflow_utils.py:
```diff
@@ -0,0 +1,97 @@
+    def save_workflow(self, **kwargs):
+        self.save_config(**kwargs)
+
+    def save_pretrained(self, save_directory: Union[str, os.PathLike], push_to_hub: bool = False, **kwargs):
```
```suggestion
    def save_pretrained(self, save_directory: Union[str, os.PathLike], push_to_hub: bool = False, filename: str = "worflow.json"):
```

patrickvonplaten commented on src/diffusers/workflow_utils.py:
```diff
@@ -0,0 +1,97 @@
+        self.save_config(**kwargs)
+
+    def save_pretrained(self, save_directory: Union[str, os.PathLike], push_to_hub: bool = False, **kwargs):
+        self.save_workflow(save_directory=save_directory, push_to_hub=push_to_hub, **kwargs)
```
```suggestion
        self.save_workflow(save_directory=save_directory, push_to_hub=push_to_hub, filename=filename)
```

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -649,6 +651,8 @@ def __call__(
                 Guidance rescale factor from [Common Diffusion Noise Schedules and Sample Steps are
                 Flawed](https://arxiv.org/pdf/2305.08891.pdf). Guidance rescale factor should fix overexposure when
                 using zero terminal SNR.
+            return_workflow(`bool`, *optional*, defaults to `False`):
```
```suggestion
            return_workflow (`bool`, *optional*, defaults to `False`):
```

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -649,6 +651,8 @@ def __call__(
                 Flawed](https://arxiv.org/pdf/2305.08891.pdf). Guidance rescale factor should fix overexposure when
                 using zero terminal SNR.
+            return_workflow(`bool`, *optional*, defaults to `False`):
+                Whether to return pipeline component configurations and call arguments.
```
```suggestion
                Whether to return used pipeline call arguments.
```

sayakpaul commented on src/diffusers/workflow_utils.py:
```diff
@@ -0,0 +1,135 @@
+        save_directory: Union[str, os.PathLike],
+        push_to_hub: bool = False,
+        filename: str = WORKFLOW_NAME,
+        **kwargs,
```
@patrickvonplaten the reason why we have it here is to support `push_to_hub()` directly on a workflow object. Otherwise, it will fail complaining about `safe_serialization` wasn't passed. 

https://github.com/huggingface/diffusers/blob/f912f39b50f87e50a9d99346f5c1b6e644653262/src/diffusers/utils/hub_utils.py#L451

sayakpaul commented:
@patrickvonplaten @DN6 could you give this a look? Once done, I will propagate the changes to the other pipelines (core ones). 

@stevhliu could you also take a look at the doc and leave your comments?

sayakpaul commented:
@Wauplin would you have any hint on why the Hub test fails? 

Locally, I am running:

```python
HUGGINGFACE_CO_STAGING=true python -m pytest -m  "is_staging_test" tests/others/test_workflows.py
```

`test_push_to_hub_in_organization()` passes just fine and I don't quite see any fundamental different in between `test_push_to_hub_in_organization()` and `test_push_to_hub()`.

sayakpaul commented on tests/others/test_workflows.py:
```diff
@@ -0,0 +1,188 @@
+@is_staging_test
+class WorkflowPushToHubTester(unittest.TestCase):
+    identifier = uuid.uuid4()
+    repo_id = f"test-workflow-{identifier}"
+    org_repo_id = f"valid_org/{repo_id}-org"
```
Thanks!

Wauplin commented on src/diffusers/workflow_utils.py:
```diff
@@ -0,0 +1,136 @@
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.config_name = WORKFLOW_NAME
+        self._internal_dict = {}
+
+    def __setitem__(self, __key, __value):
+        self._internal_dict[__key] = __value
+        return super().__setitem__(__key, __value)
+
+    def update(self, __m, **kwargs):
+        self._internal_dict.update(__m, **kwargs)
+        super().update(__m, **kwargs)
+
+    def pop(self, __key):
+        self._internal_dict.pop(__key)
+        super().pop(__key)
```
Sorry if not appropriate but I ended up reviewing the PR in the end :)

Out of curiosity, what is the purpose of `_internal_dict`? IMHO, if you inherit from `dict` it is precisely to avoid having to define the `__setitem__`, `update`, `pop`,... methods yourself. Also here a lot of magic dict methods are not implemented (`__getitem__`, `__contains__`, `get`, etc...) which creates an `_internal_dict` that will eventually end up in a corrupted state.

If `_internal_dict` has a specific purpose, I would add a comment to let future devs know about it :)

sayakpaul replied:
Thank you for your time, Lucain! Appreciate it.

Long story short, `_internal_dict` is how we maintain the configuration variables across the pipeline components. See here as an example:

https://github.com/huggingface/diffusers/blob/f912f39b50f87e50a9d99346f5c1b6e644653262/src/diffusers/configuration_utils.py#L112

Wauplin replied:
Oh ok, thanks for the explanation! So nothing to do with the fact that it's inherited from `dict` in the end. I scanned the PR looking for `_internal_dict` but did not realized it was already used in `diffusers`. All good then :+1:

sayakpaul commented:
Failing tests will go away after the changes are propagated to the pipelines. No worries.

williamberman commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -779,7 +783,20 @@ def __call__(
         # Offload all models
         self.maybe_free_model_hooks()
 
+        workflow = None
+        if return_workflow:
+            signature = inspect.signature(self.__call__)
+            argument_names = [param.name for param in signature.parameters.values()]
+            call_arg_values = inspect.getargvalues(inspect.currentframe()).locals
+            pipeline_config_name_or_path = self.config._name_or_path if hasattr(self.config, "_name_or_path") else None
+            workflow = populate_workflow_from_pipeline(argument_names, call_arg_values, pipeline_config_name_or_path)
+
```
This sort of metaprogramming can be a bit error prone e.g. if someone changes a value of a variable in the middle of the function, the return value has now changed. In my experience, this amount of introspection into the call frame/function arguments would usually just be banned. I think it's best to just be explicit here about the returned dictionary and do it immediately at the top of the call function. 

I.e. let's say we have a pipeline with some prompt cleaning function

```
def __call__(self, prompt, ..., etc...):
      prompt = self.clean_prompt(prompt)
      
      workflow = { inspect the local call frame based off of the names of the arguments to call } 

     return workflow # will return the cleaned prompt and not the actual prompt passed
```

vs this case where there are no questions about the return value
```
def __call__(self, prompt):
     workflow = {"prompt": prompt}

    prompt = self.clean_prompt(prompt)

    return workflow
```

sayakpaul replied:
Great observation!

I think one easy way to mitigate this would be to populate workflow at the beginning of `__call__()`. WDYT? If we try to be explicit, then it adds a quite a bit of verbosity which I would like to avoid.

williamberman replied:
I feel strongly that even when populated at the beginning of the method, this amount of reflection of the function arguments and inspection of the callframe shouldn't be allowed. 

Do we consider populating a dict with 10-15 values verbose compared to this? 

Populating the dict is going to be about the # of loc of values added to it for which there's zero mental overhead to read. 

Compared to ~6 loc of reflection and inspecting the callframe which has high mental overhead. 

So the trade off is an additional 4-9 loc that lets us avoid reflection and inspecting the callframe which is definitely worth it.

sayakpaul replied:
Could you elaborate more on why you feel strongly about the reflection part? Maybe provide some side-effects scenarios, so that I can understand it better. 

So, for each pipeline (which might have different call arguments), we will have to do this manually. I don't mind that but I wanted to have some automation there. 

I would appreciate opinions from @patrickvonplaten @pcuenca here as well.

stevhliu commented on docs/source/en/api/workflows.md:
````diff
@@ -0,0 +1,217 @@
+```
+
+Once the pipeline is loaded with the desired workflow, it's ready to be called:
+
````
From the Known limitation section, this might be more valuable to know up here instead of at the end.
```suggestion
<Tip>

You could also pass `prompt_embeds` instead of a `prompt`.

</Tip>
```

stevhliu reviewed (commented):
Very cool and nifty feature! 👏 I would encourage you to put this in the Guides section (maybe the Loading section as "Load a workflow") instead of the internal API section because:

1. This keeps the API section clean and dedicated as a lookup reference instead of blurring the line between APIs/Guides
2. Get more visibility and users to try out this feature
3. I think its clearly labeled "experimental" at the top so we don't have to worry too much about keeping it in the internal API section

sayakpaul commented on docs/source/en/api/workflows.md:
````diff
@@ -0,0 +1,217 @@
+image = pipeline(negative_prompt="bad quality").images[0]
+```
+
+Loading from a specific workflow is possible by specifying the `filename` argument inside the [`DiffusionPipeline.load_workflow`] method.
````
Let's say you have a repository on the Hub with a bunch of workflow files separated by different filenames. If you want to load a specific workflow by providing a filename, this feature allows you to do that.

sayakpaul commented:
@stevhliu thank you for your thoughtful suggestions. 

I think I have addressed all of them. Additionally, in https://github.com/huggingface/diffusers/pull/4825/commits/eae28adf40cd878e69a6454d2a3b9afb8723f2e1, I added some more notes to clarify the concerns around:

* Filename-specific workflows
* Whether or not a workflow can be used in a pipeline different than the one that generated it

LMKWYT.

sayakpaul commented:
@patrickvonplaten @apolinario @pcuenca 

Based on our internal Slack chat, I have:

* Added seed and device support even when no `generator` was passed. Required pipeline-level changes as you can imagine. 
* Serialized `scheduler_config` and exposed a flag to the `load_workflow()` method called `load_scheduler`, so that users can take advantage of it. I have documented it too. 

Let me know what you think. 

@stevhliu I have changed a few things in the docs again to better reflect the above and some additional things (for clarification purposes). I'd appreciate it if you could take another look.

sayakpaul commented:
@patrickvonplaten I think https://github.com/huggingface/transformers/pull/27086 is making the doc test fail. What is correct way to use `_expand_mask()` now? From [quick search](https://github.com/search?q=repo%3Ahuggingface%2Ftransformers%20_expand_mask&type=code), it seems like only TF models have the `_expand_mask()` functionality. 

I think just copying over the function here (it's only used in `pipelines/blip_diffusion/modeling_ctx_clip.py`) should work here. WDYT?

stevhliu commented on docs/source/en/using-diffusers/workflows.md:
````diff
@@ -0,0 +1,331 @@
+             ('clip_sample', False)])}
+```
+
+A [`Workflow`] object provides all the argument values in the `__call__()` of a pipeline. Add `return_workflow=True` to return a `Workflow` object:
````
I think this sentence makes more sense at the beginning of the "Serializing a workflow" section as an introduction instead of just starting with a code snippet with no context.

stevhliu commented on docs/source/en/using-diffusers/workflows.md:
````diff
@@ -0,0 +1,331 @@
+image = pipeline().images[0]
+```
+
+By default, while loading a workflow, we don't modify the scheduler of the underlying pipeline from the workflow. But you can change this behaviour by doing:
````
```suggestion
By default, while loading a workflow, the scheduler of the underlying pipeline from the workflow isn't modified but you can change it by adding `load_scheduler=True`:
```

stevhliu commented on docs/source/en/using-diffusers/workflows.md:
````diff
@@ -0,0 +1,331 @@
+pipeline.load_workflow("sayakpaul/my-simple-workflow-sd", load_scheduler=True)
+```
+
+This is particularly useful if you had changed the scheduler after loading a pipeline.
````
```suggestion
This is particularly useful if you have changed the scheduler after loading a pipeline.
```

stevhliu commented on docs/source/en/using-diffusers/workflows.md:
````diff
@@ -0,0 +1,331 @@
+image = pipeline().images[0]
+```
+
+However, make sure to thoroughly inspect the values you are calling the pipeline with, in this case.
````
Is this what you mean?

```suggestion
However, make sure the workflow values from the old pipeline are compatible with the new pipeline you're using.
```

sayakpaul replied:
Not just compatibility. If you're using a workflow that was generated using SD in SDXL, it's recommended to check the settings since the default SD height and width might not play nicely with SDXL. So, I think "inspection" is a better word here.

stevhliu commented on docs/source/en/using-diffusers/workflows.md:
````diff
@@ -0,0 +1,331 @@
+
+Loading from a local workflow is also possible:
+
+```
````
```suggestion
```py
```

stevhliu commented on docs/source/en/using-diffusers/workflows.md:
````diff
@@ -0,0 +1,331 @@
+image = pipeline().images[0]
+```
+
+Alternatively, if you want to manually load a workflow file and populate the pipeline arguments, you can do so:
````
```suggestion
Alternatively, if you want to load a workflow file and populate the pipeline arguments manually:
```

stevhliu commented on docs/source/en/using-diffusers/workflows.md:
````diff
@@ -0,0 +1,331 @@
+             ('clip_sample', False)])}
+```
+
+As you can notice, the `image` passed to the `pipeline` isn't a part of `workflow`.
````
```suggestion
```

stevhliu commented on docs/source/en/using-diffusers/workflows.md:
````diff
@@ -0,0 +1,331 @@
+workflow.save_workflow("my-simple-workflow-sd", filename="controlnet_simple.json", push_to_hub=True)
+```
+
+Steps to load the workflow into [`StableDiffusionControlNetPipeline`] should remain the same as the example above:
````
```suggestion
Then load the workflow into [`StableDiffusionControlNetPipeline`]:
```

stevhliu commented on docs/source/en/using-diffusers/workflows.md:
```diff
@@ -0,0 +1,331 @@
+
+[[autodoc]] workflow_utils.Workflow
+
+## workfllow_utils.populate_workflow_from_pipeline
```
```suggestion
## workflow_utils.populate_workflow_from_pipeline
```

stevhliu commented on docs/source/en/using-diffusers/workflows.md:
```diff
@@ -0,0 +1,331 @@
+* LoRA checkpoints: any information from LoRA checkpoints that might be loaded into a pipeline isn't serialized. Workflows generated from pipelines loaded with a LoRA checkpoint should be handled cautiously! You should ensure the LoRA checkpoint is loaded into the pipeline first before loading the corresponding workflow.
+* Call arguments including the following types: `torch.Tensor`, `np.ndarray`, `Callable`, `PIL.Image.Image`, and `List[PIL.Image.Image]`. 
+
+## Workflow
```
This could be split off into a separate `Workflow` API page in the Internal classes section :)

sayakpaul replied:
But workflows should be regular classes. I will split it off.

stevhliu commented on docs/source/en/using-diffusers/workflows.md:
```diff
@@ -0,0 +1,331 @@
+
+As you can notice, the `image` passed to the `pipeline` isn't a part of `workflow`.
+
+Let's serialize it and reload the pipeline:
```
```suggestion
Let's serialize the workflow and reload the pipeline to see what happens when you try to use it.
```

stevhliu reviewed (approved):
Thanks for the clarifications and additional info, it is clearer now! 🥳

patrickvonplaten commented on src/diffusers/pipelines/controlnet/pipeline_controlnet.py:
```diff
@@ -837,6 +843,20 @@ def __call__(
+            generator = torch.manual_seed(seed)
+
+        workflow = None
+        if return_workflow:
```
```suggestion
        if return_workflow:
                    # We do this first to capture the "true" call values. If we do this at a later point in time,
        # we cannot ensure that the true call values weren't changed during the process.
        # We update the `generator` later, though as we define a new generator in case it was passed as `None`.
        if generator is None:
            seed = random.randint(0, MAX_SEED)
            generator = torch.Generator(self.device).manual_seed(seed)
```

sayakpaul replied:
I think generator initialization should be done regardless of the status of `return_workflow`.

patrickvonplaten replied:
Why? If no workflow is returned then we don't have to know the generator no?

patrickvonplaten commented on src/diffusers/pipelines/controlnet/pipeline_controlnet.py:
```diff
@@ -837,6 +843,20 @@ def __call__(
                 second element is a list of `bool`s indicating whether the corresponding generated image contains
                 "not-safe-for-work" (nsfw) content.
         """
+        # We do this first to capture the "true" call values. If we do this at a later point in time,
+        # we cannot ensure that the true call values weren't changed during the process.
+        # We update the `generator` later, though as we define a new generator in case it was passed as `None`.
+        if generator is None:
+            seed = random.randint(0, MAX_SEED)
+            generator = torch.manual_seed(seed)
+
+        workflow = None
```
Let's try to move everything under `return_workflow:`

sayakpaul replied:
I need to understand this bit. Shouldn't we do the generator initialization first regardless of `return_workflow`? I think we should also do before `prepare_latents()` is called with `generator`, otherwise it becomes incorrect.

patrickvonplaten replied:
I don't think we need to create or measure a generator if we don't save the workflow

patrickvonplaten commented on src/diffusers/pipelines/controlnet/pipeline_controlnet.py:
```diff
@@ -1075,6 +1095,11 @@ def __call__(
-            return (image, has_nsfw_concept)
+            outputs = (image, has_nsfw_concept)
+
+            if workflow is not None:
```
```suggestion
            if return_workflow:
```

sayakpaul replied:
https://github.com/huggingface/diffusers/pull/4825#discussion_r1379577604

patrickvonplaten commented on src/diffusers/pipelines/controlnet/pipeline_controlnet.py:
```diff
@@ -837,6 +843,20 @@ def __call__(
+
+        workflow = None
+        if return_workflow:
+            signature = inspect.signature(self.__call__)
+            argument_names = [param.name for param in signature.parameters.values()]
+            call_arg_values = inspect.getargvalues(inspect.currentframe()).locals
+            workflow = populate_workflow_from_pipeline(argument_names, call_arg_values, self)
```
I'm pretty sure we can move all of this into `PipelineMixin` with a:

```py
def populate_workflow_from_pipeline(self, ....):
             signature = inspect.signature(self.__call__)
            argument_names = [param.name for param in signature.parameters.values()]
            call_arg_values = # get local from previous frame 
```

You should be able to get the local from the previous frame via: https://stackoverflow.com/questions/3711184/how-to-use-inspect-to-get-the-callers-info-from-callee-in-python

Can you maybe try that?

sayakpaul replied:
Done: https://github.com/huggingface/diffusers/pull/4825/commits/474df7a6b1629e29898ceedd173b2043877c3b4d

sayakpaul commented:
@stevhliu I decided to move the main workflows to the tutorials section and create a separate doc for the core API documentation. I hope that's okay. If not, feel free to open a follow-up PR.

patrickvonplaten commented:
We should also try to save the scheduler config as discussed on slack

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -662,6 +667,17 @@ def __call__(
+        # We do this first to capture the "True" call values. If we do this at a later point in time,
+        # we cannot ensure that the call values weren't changed during the process.
+        # We update the `generator` later, though as we define a new generator in case it was passed as `None`.
+        if generator is None:
```
Small problem here is also that this will break use cases where people make the run deterministic with `torch.manual` seed - is there a way to just retrieve which seed is currently used in torch? 

E.g. can we ensure that:

```py
torch.manual_seed(0)
image = pipe("a cat")
```

still gives exactly the same output?

patrickvonplaten commented:
Hmm sadly I think we cannot really retrieve the generator state that was used to create the latents. With the current set-up we would not necessary be able to reproduce what the user ran.

E.g. check this code:

```py
#!/usr/bin/env python3
import torch

def get_random():
    return torch.randn((2,2))

seeds = []
for _ in range(5):
    num = get_random()
    seeds.append(torch.initial_seed())
    print(num)

print("Seeds should be different, but are not", seeds)
```

I think the only option here might be to force the user to provide a `generator` when setting `return_workflow=True`

sayakpaul commented:
That works but could you elaborate how your example relates to our use case? I don't understand it.

sayakpaul commented:
@patrickvonplaten trying to address your concerns from Slack now. 

> Imagine a user who calls pipe(...) multiple times without providing a generator. In this case every time a different image will be generated, which is what we want

Here's the code I ran:

```python
from diffusers.utils import make_image_grid
from diffusers import StableDiffusionPipeline
import torch

pipeline = StableDiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, safety_checker=None
).to("cuda")

images = [pipeline("A painting of a horse", num_inference_steps=15).images[0] for _ in range(4)]
image_grid = make_image_grid(images, 2, len(images) // 2)
image_grid.save("image_grid.png")
```

I got:

![image_grid](https://github.com/huggingface/diffusers/assets/22957388/1624c1a6-31eb-4d60-b418-080dc6925f2e)

To verify:

> However torch.initial_seed() will always give you the same seed at every iteration

I tried `return_workflow=True` too:

```python
images = [pipeline("A painting of a horse", num_inference_steps=15, return_workflow=True).images[0] for _ in range(4)]
```

I got:

![image_grid_workflow](https://github.com/huggingface/diffusers/assets/22957388/a33d6675-bab4-4b03-9249-37b4f00701c8)

So, I followed your suggestion of:

> I think the only option here might be to force the user to provide a generator when setting return_workflow=True

So, now we get:

![image_grid_workflow](https://github.com/huggingface/diffusers/assets/22957388/5a360430-b2a8-43ca-aa46-8d6677291dfa)

when doing:

```python
images = [
    pipeline(
        "A painting of a horse", 
        num_inference_steps=15,
        generator=torch.manual_seed(i),
        return_workflow=True
    ).images[0] 
    for i in range(4)
]
```

Does this seem okay?

patrickvonplaten commented on src/diffusers/workflow_utils.py:
```diff
@@ -0,0 +1,91 @@
+_NON_CALL_ARGUMENTS = {"_name_or_path", "scheduler_config", "_class_name", "_diffusers_version"}
+
+
+class Workflow(dict, ConfigMixin, PushToHubMixin):
```
```suggestion
class Workflow(dict, PushToHubMixin):
```

Do we really need the ConfigMixin here, the more light-weight the better IMO

sayakpaul replied:
The reason it's there is because we can easily use `save_config()` (which is ultimately called from `save_workfllow()`). I can definitely copy over the functions being necessary to have a robust `Workflow` class but wanted to be able to reuse things here. But I see your point. Maybe having `save_workflow()` explicitly written here is better.

patrickvonplaten commented:
The following should work:

```py
#!/usr/bin/env python3
from diffusers import StableDiffusionPipeline, KDPM2DiscreteScheduler, StableDiffusionImg2ImgPipeline, HeunDiscreteScheduler, KDPM2AncestralDiscreteScheduler, DDIMScheduler
import torch
import hf_image_uploader as hiu

path = "runwayml/stable-diffusion-v1-5"

pipe = StableDiffusionPipeline.from_pretrained(path, torch_dtype=torch.float16)
pipe.enable_xformers_memory_efficient_attention()
pipe = pipe.to("cuda")

prompt = "A lion in galaxies, spirals, nebulae, stars, smoke, iridescent, intricate detail, octane render, 8k"

generator = torch.Generator(device="cpu").manual_seed(0)
# run once
outputs = pipe(prompt=prompt, generator=generator, num_images_per_prompt=1, num_inference_steps=1, return_workflow=True)
# run again with advanced generator
outputs = pipe(prompt=prompt, generator=generator, num_images_per_prompt=1, num_inference_steps=25, return_workflow=True)
image = outputs.images[0]
hiu.upload(image, "patrickvonplaten/images")

pipe.load_workflow(outputs.workflow)
image = pipe().images
hiu.upload(image, "patrickvonplaten/images")

# images should be the same
```

Getting some errors here at the moment though:

```
Traceback (most recent call last):
  File "/home/patrick/tools/./run_local.py", line 22, in <module>
    pipe.load_workflow(outputs.workflow)
  File "/home/patrick/python_bin/diffusers/pipelines/pipeline_utils.py", line 2174, in load_workflow
    device = workflow.pop("generator_device", "cpu")
TypeError: Workflow.pop() takes 2 positional arguments but 3 were given
```

Also does the workflow class really have to depend on ConfigMixin? Would prefer to keep it as light-weight as possible

sayakpaul commented:
Will have a look.

sayakpaul commented:
@patrickvonplaten there was a problem around the generator states which is why it was not generating the same results. Now, this should have been fixed. The following works:

```python
from diffusers import StableDiffusionPipeline
import numpy as np
import torch

path = "runwayml/stable-diffusion-v1-5"

pipe = StableDiffusionPipeline.from_pretrained(path, torch_dtype=torch.float16, safety_checker=None)
pipe.enable_xformers_memory_efficient_attention()
pipe = pipe.to("cuda")

prompt = "A lion in galaxies, spirals, nebulae, stars, smoke, iridescent, intricate detail, octane render, 8k"

generator = torch.Generator(device="cpu").manual_seed(0)
# run once
outputs = pipe(
    prompt=prompt, generator=generator, num_images_per_prompt=1, num_inference_steps=1, output_type="np", return_workflow=True, 
)

# run again with advanced generator
outputs = pipe(
    prompt=prompt, generator=generator, num_images_per_prompt=1, num_inference_steps=25, output_type="np", return_workflow=True
)
image_one = outputs.images[0]

pipe.load_workflow(outputs.workflow)
image_two = pipe().images[0]

image_slice_one = image_one[:3, :3, -1]
image_slice_two = image_two[:3, :3, -1]

# Evaluates to True.
print(np.allclose(image_slice_one, image_slice_two, atol=1e-4, rtol=1e-4))
```

Also, fixed some edges around setting the scheduler when `load_scheduler` is set to True in `load_workflow()`.

stevhliu commented on docs/source/en/tutorials/workflows.md:
````diff
@@ -0,0 +1,333 @@
+             ('clip_sample', False)])}
+```
+
+Once you have generated a workflow object, you can serialize it with [`~Workflow.save_workflow`]:
````
Fix link to function :)

```suggestion
Once you have generated a workflow object, you can serialize it with [`~workflow_utils.Workflow.save_workflow`]:
```

stevhliu commented on docs/source/en/tutorials/workflows.md:
````diff
@@ -0,0 +1,333 @@
+outputs.workflow.save_workflow("my-simple-workflow-sd", filename="my_workflow.json")
+```
+
+You can also set `push_to_hub=True` in [`~Workflow.save_workflow`] to directly push the workflow object to the Hub. 
````
```suggestion
You can also set `push_to_hub=True` in [`~workflow_utils.Workflow.save_workflow`] to directly push the workflow object to the Hub. 
```

stevhliu commented on docs/source/en/tutorials/workflows.md:
```diff
@@ -0,0 +1,333 @@
+
+## Serializing a workflow
+
+A [`Workflow`] object provides all the argument values in the `__call__()` of a pipeline. Add `return_workflow=True` to return a `Workflow` object. 
```
More visibility to put the sentence about the `generator` up here I think

```suggestion
A [`Workflow`] object provides all the argument values in the `__call__()` of a pipeline. Add `return_workflow=True` to return a `Workflow` object, and you *must* specify a `generator`.
```

stevhliu commented on docs/source/en/tutorials/workflows.md:
````diff
@@ -0,0 +1,333 @@
+workflow = outputs.workflow
+```
+
+<Tip warning={true}>
+
+It's mandatory to specify the `generator` when `return_workflow` is set to True.
+
+</Tip>
````
```suggestion
```

stevhliu commented on docs/source/en/tutorials/workflows.md:
```diff
@@ -0,0 +1,333 @@
+specific language governing permissions and limitations under the License.
+-->
+
+# Working with workflows
```
I think having a more descriptive title will help users discover this topic and get a better understanding of what the doc is about

```suggestion
# Share your pipeline with workflows
```

stevhliu reviewed (approved):
Hope you don't mind me adding a few more minor suggestions! :)

sayakpaul commented:
Closing this PR. When we have better bandwidth we can pick it up. So, won't delete the branch.
